### DISCUSSION TITLE
Proxy Helm Chart install through a private repo

### DISCUSSION THREAD

**Discussion by chzar** [helmfile/helmfile#1718] [Questions]
Hi!

I am a bit on the paranoid side and would like to first push a chart to a private repo before installing it in my cluster. This way, if a chart author stops publishing their chart or the repo becomes unavailable, I can still rely on my own private repository.

procedure:
1. declare the public and private repo
2. declare the private repo as a proxy for a target helm chart
3. on helmfile sync, check the private repo for the chart, if not available, fetch it from the public repo
4. install the chart in the cluster
5. if pulled from the public repo, push it to the private one.

**Comment by yxxhero** [comment]
why do you meet this user case? thanks so much.

**Reply by chzar** [reply]
Our org is concerned that if a helm chart's repo goes offline we will not have any way to deploy that chart again or have enough time to transition to something else.

**Reply by mumoshu** [reply]
@chzar Hi! Thanks for bringing this up. Interesting use-case!

All in all, I believe you'd better have a caching chart registry proxy server in front of every chart you use from helmfile, editing helmfile.yaml to point to the chart hosted in the proxy.

It could be a simple HTTP server written in, e.g., Go, which takes the proxied chart's `Cache-Control` response header into account so that it can be cached on on demand. The points 1 to 5 you mentioned would reside within the proxy, not helmfile.

That way, Helmfile can be deterministic in that it uses the chart from the proxy only, not either the proxy or the origin, easier to operate, debug, and understand.

WDYT?

**Reply by chzar** [reply]
Thats not a bad idea - could add some additional functionality (maybe as a background worker thread) to scan the manifests for container images and push them to a container registry as well. Then I would have backups of both the images and charts.

**Reply by yxxhero** [reply]
hi all, maybe we can add some hook functions.

**Reply by chzar** [reply]
I previously made a poc that used presync hooks to pull down the chart to a local file and then push it to a self hosted oci registry. The helmfile was pointed towards the the charts in the oci registry.

This did work, but made the helmfile very messy. +1 for the pull through proxy cache idea

**Reply by crabique** [reply]
Would be very cool if proxy config could be extracted into a separate Helmfile option, e.g. something like this
```YAML
repositories:
- name: argoproj
  url: https://argoproj.github.io/argo-helm
  ociCacheProxy: chartproxy.local

releases:
- name: argo-workflows
  chart: argoproj/argo-workflows
  version: 0.45.10
```

Where `chartproxy.local` would be an instance of [container-registry/helm-charts-oci-proxy](https://github.com/container-registry/helm-charts-oci-proxy) (or anything else that implements such an interface), so the chart would be actually pulled from `oci://chartproxy.local/argoproj.github.io/argo-helm/argoproj/argo-workflows` and stored there.

The two reasons it would be very nice to make this explicit are:
- this could be a higher level option (per release and could be inherited? maybe global?) so that it's easier to enforce across a big fleet of releases, centrally managed
- this would not break Renovate integration out of the box, it would just not care about the proxy config and raise an update PR as usual
